### PR TITLE
add rail-fence-cipher

### PR DESCRIPTION
**Pull request by raysapida**
I'm working on tests and a reference solution for the rail-fence-cipher exercise.

### CODE REVIEW

**Comment by raysapida** [comment]
I think the reference solution and tests are ready for review. 

For the reference solution, I wasn't sure if it was ok to use the syntax below to make the class methods besides `encode` and `decode` private. The rubocop gem was recommending another method called `private_class_method` and I could also change it to that instead. 

```ruby
class << self
    private
    ..
end
```

File: exercises/rail-fence-cipher/example.rb [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    private
+
+    def create_empty_fences_array(message, rails)
+      Array.new(rails) { Array.new(message.length) { nil } }
+    end
+
+    def populate_fences_array(message, rails, fences)
+      row = 0
+      row_direction = 1
+      (0..(message.length - 1)).each do |column|
```
Comment (bernardoamc) [inline exercises/rail-fence-cipher/example.rb]: We could use `...` here, meaning that you want to stop right before the `message.length`.

Another possibility is to use the `times` method, in this case it would be  `message.length.times do |column|`

File: exercises/rail-fence-cipher/example.rb [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    lines = split_the_encrypted_message_by_rows(message, line_lengths)
+    result = switch_marked_cells_with_string(mark_fences, lines, empty)
+    result.transpose.flatten.join
+  end
+
+  class << self
+    private
+
+    def create_empty_fences_array(message, rails)
+      Array.new(rails) { Array.new(message.length) { nil } }
```
Comment (bernardoamc) [inline exercises/rail-fence-cipher/example.rb]: The default of `Array.new(number)` is generating an array with nils, so we could also do: 
`Array.new(rails) { Array.new(message.length) }

File: exercises/rail-fence-cipher/rail_fence_cipher_test.rb [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    assert_equal 'WECRLTEERDSOEEFEAOCAIVDEN',
+                 RailFenceCipher.encode('WEAREDISCOVEREDFLEEATONCE', 3)
+  end
+
+  def test_encode_with_ending_in_the_middle
+    skip
+    assert_equal 'ESXIEECSR', RailFenceCipher.encode('EXERCISES', 4)
+  end
+
+  def test_encode_with_a_more_diverse_text
```
Comment (bernardoamc) [inline exercises/rail-fence-cipher/rail_fence_cipher_test.rb]: I think this test is not really exercising a new feature or edge case, could this be removed?

File: exercises/rail-fence-cipher/rail_fence_cipher_test.rb [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+                 RailFenceCipher.decode('XXXXXXXXXOOOOOOOOO', 2)
+  end
+
+  def test_decode_with_three_rails
+    skip
+    assert_equal 'THEDEVILISINTHEDETAILS',
+                 RailFenceCipher.decode('TEITELHDVLSNHDTISEIIEA', 3)
+  end
+
+  def test_decode_with_a_more_diverse_text
```
Comment (bernardoamc) [inline exercises/rail-fence-cipher/rail_fence_cipher_test.rb]: I don't know exactly if adding a new rail exercises a new aspect of the code, if it doesn't we could trim a few test cases. What do you think?

File: exercises/rail-fence-cipher/rail_fence_cipher_test.rb [hunk 5]
```diff
@@ -0,0 +1,94 @@
+gem 'minitest', '>= 5.0.0'
+require 'minitest/autorun'
+require_relative 'rail_fence_cipher'
+
+# rubocop:enable all
+class RailFenceCipherTest < Minitest::Test
+  def test_encode_with_empty_string
+    assert_equal '', RailFenceCipher.encode('', 4)
```
Comment (bernardoamc) [inline exercises/rail-fence-cipher/rail_fence_cipher_test.rb]: I like the use of a class method here, this gives people the opportunity to use a `module` to solve the exercise if they want to. 👍

**Comment by bernardoamc** [comment]
Nice @raysapida ! 🎉 

I think it would be good to use `private_class_method`, but this is just my preference! :)

File: exercises/rail-fence-cipher/example.rb (hunk 1 above)
Comment (raysapida) [inline exercises/rail-fence-cipher/example.rb]: For me, it looks like `message.length.times` would be the most readable. Thank you for suggesting it!

File: exercises/rail-fence-cipher/example.rb (hunk 2 above)
Comment (raysapida) [inline exercises/rail-fence-cipher/example.rb]: Ohh , I see. Would it be better to be explicit that it's populating with nil or is it a common assumption that the default is nil?

File: exercises/rail-fence-cipher/rail_fence_cipher_test.rb (hunk 3 above)
Comment (raysapida) [inline exercises/rail-fence-cipher/rail_fence_cipher_test.rb]: I think it could be removed. I was adding most of the tests cases from different implementations and they do have a lot of overlap. In those situations, would it be better to have just enough tests to cover the feature and edge cases?

File: exercises/rail-fence-cipher/rail_fence_cipher_test.rb (hunk 4 above)
Comment (raysapida) [inline exercises/rail-fence-cipher/rail_fence_cipher_test.rb]: While I was going through the challenge, it seemed like adding a new rail helped me make sure that I wasn't depending on the wrong pattern.  I first tried to populate the fences without marking them and I found out that the logic I was using wouldn't work past the third rail.

File: exercises/rail-fence-cipher/example.rb (hunk 2 above)
Comment (bernardoamc) [inline exercises/rail-fence-cipher/example.rb]: That's a good question, I think we could say it is a common assumption. :)

File: exercises/rail-fence-cipher/rail_fence_cipher_test.rb (hunk 3 above)
Comment (bernardoamc) [inline exercises/rail-fence-cipher/rail_fence_cipher_test.rb]: I think so, when I see a test I always wonder what feature/edge case it is covering. If it is not doing this I usually remove it since it is more code for people to read that is not helping them to understand the logic.

File: exercises/rail-fence-cipher/rail_fence_cipher_test.rb (hunk 4 above)
Comment (bernardoamc) [inline exercises/rail-fence-cipher/rail_fence_cipher_test.rb]: Yeah, tests are awesome during development, but you have found that the same logic applies to all cases now, right? Maybe this could even keep people wondering if they have to implement something different for texts with 4 or more fences, for example.

File: exercises/rail-fence-cipher/rail_fence_cipher_test.rb (hunk 3 above)
Comment (kotp) [inline exercises/rail-fence-cipher/rail_fence_cipher_test.rb]: A name such as "a more diverse..." that can not be clear in what it is testing, is likely a hint it is not necessary, as it can not be articulated what specifically it is testing for.

File: exercises/rail-fence-cipher/rail_fence_cipher_test.rb (hunk 4 above)
Comment (raysapida) [inline exercises/rail-fence-cipher/rail_fence_cipher_test.rb]: Yeah, it does all apply to all of those cases now. And I could see what you meant earlier that having a lot of tests wouldn't necessarily help make the code easier to understand if they're covering the same things.

File: exercises/rail-fence-cipher/rail_fence_cipher_test.rb (hunk 5 above)
Comment (kotp) [inline exercises/rail-fence-cipher/rail_fence_cipher_test.rb]: Interestingly, this doesn't only give people the opportunity to use a module, but it does free up the solution to be pretty open.

**Comment by kotp** [comment]
Looks good to me, I think this is ready for merge.

**Comment by kytrinyx** [comment]
I agree. @raysapida would you rebase this onto master to resolve the conflicts? (We can also do that for you, just let us know).

**Comment by raysapida** [comment]
Hi @kytrinyx and @kotp . Would rebasing it onto master be similar to this?

**Comment by kotp** [comment]
@raysapida That would be the way to do it. :cake:
